File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/YoBit.java
package com.mobnetic.coinguardian.model.market;

import java.util.List;
import java.util.Locale;

import org.json.JSONArray;
import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardiandatamodule.R;

public class YoBit extends Market {

	private final static String NAME = "YoBit";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://yobit.net/api/3/ticker/%1$s";
	private final static String URL_CURRENCY_PAIRS = "https://yobit.net/api/3/info";
	
	public YoBit() {
		super(NAME, TTS_NAME, null);
	}
	
	@Override
	public int getCautionResId() {
		return R.string.market_caution_yobit;
	}
	
	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyPairId());
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		final JSONArray names = jsonObject.names();
		JSONObject tickerJsonObject = jsonObject.getJSONObject(names.getString(0));
		
		ticker.bid = tickerJsonObject.getDouble("sell");
		ticker.ask = tickerJsonObject.getDouble("buy");
		ticker.vol = tickerJsonObject.getDouble("vol");
		ticker.high = tickerJsonObject.getDouble("high");
		ticker.low = tickerJsonObject.getDouble("low");
		ticker.last = tickerJsonObject.getDouble("last");
		ticker.timestamp = tickerJsonObject.getLong("updated");
	}
	
	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL_CURRENCY_PAIRS;
	}
	
	@Override
	protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONObject pairsJsonObject = jsonObject.getJSONObject("pairs");
		final JSONArray pairsNames = pairsJsonObject.names();
		for(int i=0; i<pairsNames.length(); ++i) {
			String pairId = pairsNames.getString(i);
			if(pairId==null)
				continue;
			String[] currencies = pairId.split("_");
			if(currencies.length!=2)
				continue;
			
			String currencyBase = currencies[0].toUpperCase(Locale.ENGLISH);
			String currencyCounter = currencies[1].toUpperCase(Locale.ENGLISH);
			pairs.add(new CurrencyPairInfo(currencyBase, currencyCounter, pairId));
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/YoBit.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.R
import com.aneonex.bitcoinchecker.datamodule.util.forEachString
import org.json.JSONObject
import java.util.*

class YoBit : Market(NAME, TTS_NAME, null) {
    companion object {
        private const val NAME = "YoBit"
        private const val TTS_NAME = NAME
        private const val URL = "https://yobit.net/api/3/ticker/%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://yobit.net/api/3/info"
    }

    override val cautionResId: Int
        get() = R.string.market_caution_yobit

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, checkerInfo.currencyPairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        val names = jsonObject.names()!!
        val tickerJsonObject = jsonObject.getJSONObject(names.getString(0))
        ticker.bid = tickerJsonObject.getDouble("sell")
        ticker.ask = tickerJsonObject.getDouble("buy")
        ticker.vol = tickerJsonObject.getDouble("vol")
        ticker.high = tickerJsonObject.getDouble("high")
        ticker.low = tickerJsonObject.getDouble("low")
        ticker.last = tickerJsonObject.getDouble("last")
        ticker.timestamp = tickerJsonObject.getLong("updated")
    }

    // ====================
    // Get currency pairs
    // ====================
    override fun getCurrencyPairsUrl(requestId: Int): String {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        jsonObject.getJSONObject("pairs").names()!!.forEachString { pairId ->
            val currencies = pairId.split('_')
            if (currencies.size == 2) {
                val currencyBase = currencies[0].uppercase(Locale.ROOT)
                val currencyCounter = currencies[1].uppercase(Locale.ROOT)
                pairs.add(CurrencyPairInfo(currencyBase, currencyCounter, pairId))
            }
        }
    }
}